### PR TITLE
Refactor interface instance typechecking and compilation (scala)

### PR DESCRIPTION
**Pull request by akrmn**
This PR seeks to reduce duplication between `Implements` and `CoImplements` so we can treat them as uniformly as possible

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -694,41 +700,38 @@ private[lf] final class Compiler(
...
 
-  // Witness the fact that the template 'tmplId' implements the interface 'ifaceId'
-  private[this] def compileImplements(
-      tmplId: Identifier,
-      ifaceId: Identifier,
-  ): (t.SDefinitionRef, SDefinition) =
-    t.ImplementsDefRef(tmplId, ifaceId) -> UnitDef
+  // Compile the contents of an interface instance, including a witness for
+  // the existence of said interface instance.
+  private[this] def compileInterfaceInstance(
```
Comment (akrmn) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:705]: Replaces `compileImplements`+`foreach(compileImplementsMethod)` (resp. `compileCoImplements`+`foreach(compileCoImplementsMethod)`) for `TemplateImplements` (resp. `InterfaceCoImplements`)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 2]
```diff
@@ -694,41 +700,38 @@ private[lf] final class Compiler(
...
+    // TODO https://github.com/digital-asset/daml/issues/14112
+    // compile interface instance view
+
+    builder.result()
   }
 
-  // Compile the interface-provided implementation of a method for the given template.
-  private[this] def compileCoImplementsMethod(
+  // Compile the implementation of an interface method.
+  private[this] def compileInterfaceInstanceMethod(
```
Comment (akrmn) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:729]: replaces `compileImplementsMethod` and `compileCoImplementsMethod`

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 3]
```diff
@@ -441,46 +441,36 @@ object SExpr {
...
+    * points to the Unit value if 'parent' defines an interface instance
+    * of the interface for the template.
+    *
+    * invariants:
+    *   * parent == interfaceId || parent == templateId
+    *   * at most one of the following is defined:
+    *       InterfaceInstanceDefRef(i, i, t)
+    *       InterfaceInstanceDefRef(t, i, t)
+    *
+    * The parent is used to determine what package and module define
```
Comment (akrmn) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:454]: @cocreature @remyhaemmerle-da, this is the reason we need the `parent` (previously, the reason we needed to override `ref`)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 4]
```diff
@@ -441,46 +441,36 @@ object SExpr {
...
+    * the interface instance, which is used to fetch the appropriate
+    * package in case it's missing.
```
Comment (akrmn) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:456]: @cocreature @remyhaemmerle-da, this is the reason we need the `parent` (previously, the reason we needed to override `ref`)

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/LookupError.scala [hunk 5]
```diff
@@ -103,23 +104,35 @@ object Reference {
...
-  }
-
-  final case class TemplateImplementsOrInterfaceCoImplements(
-      templateName: TypeConName,
-      ifaceName: TypeConName,
+  /** References an interface implementation of interfaceName for templateName
+    * defined in templateName if parentTemplateOrInterface == TorI.Template(())
+    * or in interfaceName if parentTemplateOrInterface == TorI.Interface(())
+    */
+  final case class ConcreteInterfaceInstance(
```
Comment (akrmn) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/LookupError.scala:119]: this one is only used as a `Context` in `Typing.scala` for better error messages. We don't use it in `lookupX` functions since we never look explicitly for an "interface instance in a template" or an "interface instance in an interface"

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/PackageInterface.scala [hunk 6]
```diff
@@ -214,82 +214,44 @@ private[lf] class PackageInterface(signatures: PartialFunction[PackageId, Packag
...
+      interfaceName: TypeConName,
+      templateName: TypeConName,
       context: => Reference,
   ): Either[
-    LookupError,
-    TemplateOrInterface[
-      TemplateImplementsSignature,
-      InterfaceCoImplementsSignature,
-    ],
+    Either[LookupError, AmbiguousInterfaceInstanceError],
```
Comment (akrmn) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/PackageInterface.scala:222]: the nested `Either` is because we have two different error conditions: no `interface instance` exists at all, or both possible ones do. We could have folded `AmbiguousInterfaceInstanceError` into `LookupError`, but quite a few places match on `LookupError` so that would affect too much unrelated code.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala [hunk 7]
```diff
@@ -378,43 +371,38 @@ final case class EModuleVersionDependencies(
...
-final case class EMissingInterfaceMethod(
+final case class EMissingMethodInInterfaceInstance(
     context: Context,
-    template: TypeConName,
-    iface: TypeConName,
     method: MethodName,
 ) extends ValidationError {
   override protected def prettyInternal: String =
-    s"Template $template is missing method '$method' in its implementation of interface $iface."
+    s"Interface instance lacks an implementation for method '$method'."
```
Comment (akrmn) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/ValidationError.scala:379]: This takes fewer arguments because all the relevant information is in `context` and `ValidationError.pretty` prepends that information to `prettyInternal`.

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM. Thanks
